src/sentry/notifications/helpers.py

```diff
@@ -870,89 +827,152 @@ def get_setting_options_for_users(
+        additional_filters: Additional filters to apply to the query for NotificationSettingProvider.
     """
-    users = user_service.get_many(filter={"user_ids": user_ids})
-    query = get_query(recipients=users, project=project, organization=organization)
-    notification_settings = NotificationSettingProvider.objects.filter(query)
+    notification_settings = get_all_setting_providers(
+        recipients, projects, organization, additional_filters
+    )
 
+    # Generate a map of user to the most specific setting for each provdier for each notification type.
     user_to_setting: MutableMapping[
-        RpcActor, MutableMapping[ExternalProviderEnum, NotificationSettingProvider]
+        RpcActor,
+        MutableMapping[
+            NotificationSettingEnum,
+            MutableMapping[ExternalProviderEnum, NotificationSettingProvider],
+        ],
     ] = defaultdict(dict)
     for ns in notification_settings:
         ns_dict = user_to_setting[ns.user]
+        setting_type = NotificationSettingEnum(ns.type)
         provider = ExternalProviderEnum(ns.provider)
+        if setting_type not in ns_dict:
+            ns_dict[setting_type] = {}
+
         if ns.scope_type == NotificationScopeEnum.PROJECT.value:
             # Project settings take precedence, so we can overwrite any existing settings.
-            ns_dict[provider] = ns
-
+            ns_dict[setting_type].update({provider: ns})
         if ns.scope_type == NotificationScopeEnum.ORGANIZATION.value:
-            if provider not in ns_dict:
-                ns_dict[provider] = ns
+            if setting_type not in ns_dict or provider not in ns_dict[setting_type]:
+                ns_dict[setting_type].update({provider: ns})
             else:
                 # Organization settings do not override project settings
-                if ns_dict[provider].scope_type != NotificationScopeEnum.PROJECT.value:
-                    ns_dict[provider] = ns
-
+                if (
+                    ns_dict[setting_type][provider].scope_type
+                    != NotificationScopeEnum.PROJECT.value
+                ):
+                    ns_dict[setting_type].update({provider: ns})
         if ns.scope_type == NotificationScopeEnum.USER.value:
-            if provider not in ns_dict:
-                ns_dict[provider] = ns
+            if setting_type not in ns_dict or provider not in ns_dict[setting_type].keys():
+                ns_dict[setting_type].update({provider: ns})
             else:
                 # User settings do not override project or organization settings
-                if ns_dict[provider].scope_type not in [
+                if ns_dict[setting_type][provider].scope_type not in [
                     NotificationScopeEnum.PROJECT.value,
                     NotificationScopeEnum.ORGANIZATION.value,
                 ]:
-                    ns_dict[provider] = ns
+                    ns_dict[setting_type].update({provider: ns})
```

Comment:
it seems like these lines within these ifs (other than 890) don't have test coverage -- should we add some?